File: plugins/plugin-nostr/rust/src/actions/publish_profile.rs
//! Publish profile action for the Nostr plugin.

use crate::service::NostrService;
use crate::types::*;
use serde::{Deserialize, Serialize};
use tracing::debug;

/// Parameters for publishing a Nostr profile
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublishProfileParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub about: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub picture: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub banner: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nip05: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lud16: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub website: Option<String>,
}

/// Result of the publish profile action
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublishProfileResult {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event_id: Option<String>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub relays: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Execute the publish profile action
pub async fn execute_publish_profile(
    service: &NostrService,
    params: PublishProfileParams,
) -> PublishProfileResult {
    // Check connection
    if !service.is_connected().await {
        return PublishProfileResult {
            success: false,
            event_id: None,
            relays: Vec::new(),
            error: Some("Nostr service not connected".to_string()),
        };
    }

    // Create profile
    let profile = NostrProfile {
        name: params.name,
        display_name: params.display_name,
        about: params.about,
        picture: params.picture,
        banner: params.banner,
        nip05: params.nip05,
        lud16: params.lud16,
        website: params.website,
    };

    // Publish profile
    let result = service.publish_profile(profile).await;

    if !result.success {
        return PublishProfileResult {
            success: false,
            event_id: None,
            relays: Vec::new(),
            error: result.error,
        };
    }

    debug!("Published profile, event_id: {:?}", result.event_id);

    PublishProfileResult {
        success: true,
        event_id: result.event_id,
        relays: result.relays,
        error: None,
    }
}

/// Action metadata
pub const PUBLISH_PROFILE_ACTION_NAME: &str = "NOSTR_PUBLISH_PROFILE";
pub const PUBLISH_PROFILE_ACTION_DESCRIPTION: &str =
    "Publish or update the bot's Nostr profile (kind:0 metadata)";
pub const PUBLISH_PROFILE_ACTION_SIMILES: &[&str] = &[
    "UPDATE_NOSTR_PROFILE",
    "SET_NOSTR_PROFILE",
    "NOSTR_PROFILE",
];

File: plugins/plugin-nostr/typescript/src/actions/publishProfile.ts
/**
 * Publish profile action for Nostr plugin.
 */

import {
  type Action,
  type ActionResult,
  composePromptFromState,
  type IAgentRuntime,
  type Memory,
  ModelType,
  parseJSONObjectFromText,
  type State,
} from "@elizaos/core";
import type { NostrService } from "../service.js";
import { NOSTR_SERVICE_NAME, type NostrProfile } from "../types.js";

const PUBLISH_PROFILE_TEMPLATE = `# Task: Extract Nostr profile data
Based on the conversation, determine what profile information to update.

Recent conversation:
{{recentMessages}}

Extract any of the following profile fields that should be updated:
- name: Display name
- about: Bio/description
- picture: Profile picture URL
- banner: Banner image URL
- nip05: Nostr verification (user@domain.com)
- lud16: Lightning address (user@domain.com)
- website: Website URL

Respond with a JSON object containing only the fields to update:
\`\`\`json
{
  "name": "optional name",
  "about": "optional bio"
}
\`\`\``;

export const publishProfile: Action = {
  name: "NOSTR_PUBLISH_PROFILE",
  similes: ["UPDATE_NOSTR_PROFILE", "SET_NOSTR_PROFILE", "NOSTR_PROFILE"],
  description: "Publish or update the bot's Nostr profile (kind:0 metadata)",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
  ): Promise<boolean> => {
    return message.content.source === "nostr";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: Record<string, unknown>,
    callback?: (response: { text: string; source?: string }) => void,
  ): Promise<ActionResult> => {
    const nostrService = runtime.getService<NostrService>(NOSTR_SERVICE_NAME);

    if (!nostrService || !nostrService.isConnected()) {
      if (callback) {
        callback({ text: "Nostr service is not available.", source: "nostr" });
      }
      return { success: false, error: "Nostr service not available" };
    }

    // Get or compose state
    const currentState = state ?? (await runtime.composeState(message));

    // Compose prompt
    const prompt = await composePromptFromState({
      template: PUBLISH_PROFILE_TEMPLATE,
      state: currentState,
    });

    // Extract parameters using LLM
    let profileInfo: NostrProfile | null = null;
    for (let attempt = 0; attempt < 3; attempt++) {
      const response = await runtime.useModel(ModelType.TEXT_SMALL, {
        prompt,
      });

      const parsed = parseJSONObjectFromText(String(response));
      if (parsed) {
        profileInfo = {
          name: parsed.name ? String(parsed.name) : undefined,
          displayName: parsed.displayName
            ? String(parsed.displayName)
            : undefined,
          about: parsed.about ? String(parsed.about) : undefined,
          picture: parsed.picture ? String(parsed.picture) : undefined,
          banner: parsed.banner ? String(parsed.banner) : undefined,
          nip05: parsed.nip05 ? String(parsed.nip05) : undefined,
          lud16: parsed.lud16 ? String(parsed.lud16) : undefined,
          website: parsed.website ? String(parsed.website) : undefined,
        };
        break;
      }
    }

    if (!profileInfo) {
      if (callback) {
        callback({
          text: "I couldn't understand the profile information. Please try again.",
          source: "nostr",
        });
      }
      return { success: false, error: "Could not extract profile parameters" };
    }

    // Publish profile
    const result = await nostrService.publishProfile(profileInfo);

    if (!result.success) {
      if (callback) {
        callback({
          text: `Failed to publish profile: ${result.error}`,
          source: "nostr",
        });
      }
      return { success: false, error: result.error };
    }

    if (callback) {
      callback({
        text: "Profile published successfully.",
        source: message.content.source as string,
      });
    }

    return {
      success: true,
      data: {
        eventId: result.eventId,
        relays: result.relays,
        profile: profileInfo,
      },
    };
  },

  examples: [
    [
      {
        name: "{{user1}}",
        content: { text: "Update your profile name to 'Bot Assistant'" },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll update my Nostr profile.",
          actions: ["NOSTR_PUBLISH_PROFILE"],
        },
      },
    ],
  ],
};
